x/pairing/keeper/unresponsive_provider.go

```diff
@@ -111,84 +109,118 @@ func (k Keeper) getUnresponsiveProvidersToPunish(ctx sdk.Context, currentEpoch u
+			// Get previous epoch (from epochTemp)
+			previousEpoch, err := k.epochStorageKeeper.GetPreviousEpochStartForBlock(ctx, epochTemp)
+			if err != nil {
+				return utils.LavaError(ctx, k.Logger(ctx), "get_previous_epoch", map[string]string{"err": err.Error(), "epoch": fmt.Sprintf("%+v", epochTemp)}, "couldn't get previous epoch")
+			}
+			// update epochTemp
+			epochTemp = previousEpoch
+
+			// there is no providerPaymentStorage for this provider in this epoch -> continue searching
+			continue
 		}
 
-		// Go over providerPaymentStorage list
-		for _, providerPaymentStorage := range providerPaymentStorageList {
+		// counter is smaller than epochsNumToCheckCUForUnresponsiveProvider -> only count CU serviced by the provider in the epoch
+		if counter < epochsNumToCheckCUForUnresponsiveProvider {
+			// count the CU by iterating through the uniquePaymentStorageClientProvider objects
 			cuCount := uint64(0)
+			for _, uniquePayment := range providerPaymentStorage.GetUniquePaymentStorageClientProvider() {
+				cuCount += uniquePayment.GetUsedCU()
+			}
 
-			// Get provider address and chainID
-			providerAddress := k.getProviderAddressFromProviderPaymentStorageKey(providerPaymentStorage.Index)
-			chainID := k.getChainIDFromProviderPaymentStorageKey(providerPaymentStorage.Index)
-
-			// Create a punishEntryKey from the provider's address and the chain ID
-			punishEntryKey := providerAddress + "_" + chainID
-
-			// find the provider's punishEntry in map
-			providerPunishEntry, punishEntryFound := providersPunishEntryMap[punishEntryKey]
-
-			if isProvider {
-				// if isProvider is true, we count the CU serviced by the unersponsive provider
-				for _, uniquePayment := range providerPaymentStorage.GetUniquePaymentStorageClientProvider() {
-					cuCount += uniquePayment.GetUsedCU()
-				}
-				if punishEntryFound {
-					// punishEntry found -> add the cuCount to providerServicedCu and update the map entry
-					providerPunishEntry.providerServicedCu += cuCount
-					providersPunishEntryMap[punishEntryKey] = providerPunishEntry
-				} else {
-					// punishEntry not found -> create a new one and add it to the map
-					providersPunishEntryMap[punishEntryKey] = punishEntry{complainersUsedCu: 0, providerServicedCu: cuCount, providerPaymentStorageIndexList: []string{providerPaymentStorage.Index}}
-				}
+			// update the CU in the providerCuCounterForUnreponsivenessMap
+			if createNewproviderCuCounterForUnreponsivenessEntryFlag {
+				// providerCuCounterForUnreponsiveness not found -> create a new one and add it to the map
+				providerCuCounterForUnreponsivenessMap[providerCuCounterForUnreponsivenessKey] = providerCuCounterForUnreponsiveness{complainersUsedCu: 0, providerServicedCu: cuCount, providerPaymentStorageIndexList: []string{providerPaymentStorage.GetIndex()}}
+
+				// set the createNewproviderCuCounterForUnreponsivenessEntryFlag to false
+				createNewproviderCuCounterForUnreponsivenessEntryFlag = false
 			} else {
```

Comment:
this isnt suppose to be an else, the first epochs you count for both complainers and providers, do if counter < epochsComplainers instead

Reply:
this was what I did in the previous implementation that you told me to fix (the fix was go back 4 and then 2, not counting simultaneously)

Reply:
done